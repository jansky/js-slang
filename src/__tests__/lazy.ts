import {
  expectResult,
} from '../utils/testing'

test('Unused arguments are not evaluated', () => {

  return expectResult('function test(a, b) { return a === 1 ? a : b; } test(1, head(null));',
    {executionMethod: 'interpreter_lazy', chapter: 2}).toBe(1);

});

test('Unary operations force argument', () => {

  return expectResult('function neg(b) { return !b; } neg(((x) => x)(false)); ',
    {executionMethod: 'interpreter_lazy'}).toBe(true);

});

test('Binary operations force arguments', () => {

  return expectResult('function add(x, y) { return x + y; } add(((x) => x)(5), ((x) => x + 1)(9)); ',
    {executionMethod: 'interpreter_lazy'}).toBe(15);

});

test('Conditionals force test', () => {

  return expectResult(`
function f(a, b) {
  return (a ? true : head(null)) && (!b ? true : head(null));
}

f(((b) => b)(true), ((b) => !b)(true));
`,    {executionMethod: 'interpreter_lazy', chapter: 2}).toBe(true);

});

test('Thunks are memoized', () => {

  return expectResult(`
let x = 1;

function incX() {
  x = x + 1;
  return x;
}

function square(n) {
  return n * n;
}

square(incX());`, {executionMethod: 'interpreter_lazy', chapter: 3}).toBe(4);

});