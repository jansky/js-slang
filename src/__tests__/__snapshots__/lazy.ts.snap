// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Binary operations force arguments: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function add(x, y) { return x + y; } add(((x) => x)(5), ((x) => x + 1)(9)); ",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 15,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Conditionals force test: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
function f(a, b) {
  return (a ? true : head(null)) && (!b ? true : head(null));
}

f(((b) => b)(true), ((b) => !b)(true));
",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Thunks are memoized: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
let x = 1;

function incX() {
  x = x + 1;
  return x;
}

function square(n) {
  return n * n;
}

square(incX());",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 4,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Unary operations force argument: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function neg(b) { return !b; } neg(((x) => x)(false)); ",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Unused arguments are not evaluated: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function test(a, b) { return a === 1 ? a : b; } test(1, head(null));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
